portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,33 +14,95 @@
 public class DataServlet extends HttpServlet {
-  private List<String> commentsStorage = new ArrayList<>();
+  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String json = new Gson().toJson(commentsStorage);
+    Query query = new Query("Comment");
+    Iterable<Entity> results = (datastore.prepare(query)).asIterable();
+    List<String> commentsStorage = new ArrayList<>();
+    List<Integer> boundsStorage = new ArrayList<>();
+
+    // Collect all comments and the preferred maximum numbers of comments.
+    results.forEach(e -> {
+      addCommentContents(e, commentsStorage);
+      getMaxCommentsNum(e, boundsStorage);
+    });
+
+    // Show the specific number of comments.
+    String json = new Gson().toJson(collectShownComments(commentsStorage, boundsStorage));
     response.setContentType("text/html");
     response.getWriter().println(json);
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String comment = request.getParameter("comments-input");
-    commentsStorage.add(comment);
+    String maxCommentsNumString = request.getParameter("max-comments-num");
+    int maxCommentsNum = 0;
+    if (maxCommentsNumString != null) {
+      maxCommentsNum = Integer.parseInt(maxCommentsNumString);
+    }
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("contents", comment);
+    commentEntity.setProperty("bound", maxCommentsNum);
+    datastore.put(commentEntity);
     response.setContentType("text/html;");
     response.getWriter().println("Thank you for your comments!");
-    response.sendRedirect("/index.html");
+  }
+
+  /** For a specific entity, add the inserted maximum comments' number to the collection. */
+  public void getMaxCommentsNum(Entity entity, List<Integer> boundsStorage) {
+    if (entity.getProperty("bound") != null) {
+      long boundLong = (long) entity.getProperty("bound");
+      int bound = Math.toIntExact(boundLong);
+      boundsStorage.add(bound);
+    }
+  }
+
+  /** For a specific entity, add its comment's contents to the storage. */
+  public void addCommentContents(Entity entity, List<String> commentsStorage) {
+    String contents = (String) entity.getProperty("contents");
+    if (contents != "") {
+      commentsStorage.add(contents);
+    }
+  }
+
+  /** Controls the output comments' collection with the maximum comments' number limitation. */
+  public List<String> collectShownComments(
+      List<String> commentsStorage, List<Integer> boundsStorage) {
+    // If nobody inserts a preferred maximum comments' number, the number is set to 3 as default.
+    int maxCommentsNum = 3;
+
+    // Otherwise, the number is the maximum value of all inserted numbers.
+    if (!boundsStorage.isEmpty()) {
+      maxCommentsNum = Collections.max(boundsStorage);
+    }
```

Comment:
Could this be extracted to a method like calculateNumberOfCommentsToReturn ?

Reply:
Sure! I have further broken this helper function into two - "collectShownComments" and "getMaxCommentNum".